RFC: Fix URTS (#26)
iotaledger/protocol-rfcs: text/0008-uniform-random-tip-selection/0008-uniform-random-tip-selection.md

+ Feature name: `uniform-random-tip-selection`
+ Start date: 2020-03-09
+ RFC PR: [iotaledger/protocol-rfcs#0008](https://github.com/iotaledger/protocol-rfcs/pull/0008)

> # Summary
> 
> Weighted Uniform Random Tip Selection on a subset enables a node to perform fast tip-selection to increase message throughput.
> The algorithm selects tips which are non-lazy in order to maximize confirmation rate.
> 
> # Motivation
> 
> Because of the `white-flag` confirmation algorithm, it is no longer necessary to perform complex
> tip-selection which evaluates ledger mutations while walking, therefore, a simpler and better 
> performing algorithm can be used to select tips, which in turn increases overall message throughput.
> 
> In order to maximize confirmation rate however, the algorithm needs to return tips which are `non-lazy`.
> Non-lazy in this context means that a tip does not attach to a cone of messages which is too far
> in the past. Such a cone is likely to be already confirmed and does not contribute to the
> rate of newly confirmed messages when a milestone is issued.

Reviewer: ```suggestion
# Summary

Weighted Uniform Random Tip Selection on a subset enables a node to perform fast tip-selection to increase message throughput.
The algorithm selects tips which are non-lazy to maximize confirmation rate.

# Motivation

Because of the `white-flag` confirmation algorithm, it is no longer necessary to perform complex
tip-selection which evaluates ledger mutations while walking. Therefore, a more simple, better 
performing algorithm can be used to select tips, which in turn increases overall message throughput.

To maximize confirmation rate however, the algorithm needs to return tips which are `non-lazy`.
Non-lazy in this context means that a tip does not attach to a cone of messages which is too far
in the past. Such a cone is likely to be already confirmed and does not contribute to the
rate of newly confirmed messages when a milestone is issued.
```

when we walk the past cone of a given message. The walk stops on a confirmed message.  
Note that the red marked milestone is also a `Confirmed Root Message`.
![sdf](images/cnf_tx_roots.PNG)
* `Message Snapshot Index (MSI)` defines the index of the milestone which confirmed a given message.
* `Oldest Message Root Snapshot Index (OMRSI)` defines the lowest milestone index of a set of
`Confirmed Root Messages` of a given messages.
* `Youngest Message Root Snapshot Index (YMRSI)` defines the highest milestone index of a set of
`Confirmed Root Messages` of a given message.
* `Latest Solid Milestone Index (LSMI)` the index of the latest solid milestone.
* `Below Max Depth (BMD)` defines a threshold value up on which it is decided on whether a message is not
relevant in relation to the recent parts of the Tangle. The current `BMD` for mainnet nodes is 15 milestones, 
which means that messages of which their `OMRSI` in relation to the `LSMI` is more than 15, are "below max depth".

### OMRSI / YMRSI example
Given the blue PoV message, the `OMRSI` of it is milestone 1 and `YMRSI` milestone 2.

> Note that here again, the milestones are also `Confirmed Root Messages`.

Reviewer: ```suggestion
Note that, here again, the milestones are also `Confirmed Root Messages`.
```

* `1`: The tip is somewhat lazy.
* `2`: The tip is a non-lazy tip.

Definitions:
* `C1`: Max allowed delta value for the `YMRSI` of a given message in relation to the current `LSMI`.
* `C2`: Max allowed delta value between `OMRSI` of a given message in relation to the current `LSMI`. 
* `M`: Max allowed delta value between `OMRSI` of the given message in relation to the current `LSMI`.
`M` is the `below max depth (BMD)` parameter.

Recommended defaults:
* `C1` = 8 milestones
* `C2` = 13 milestones
* `M` = 15 milestones

Scoring Algorithm (pseudo code):
```

enum Score (
    LAZY = 0
    SEMI_LAZY = 1
    NON_LAZY = 2
)

const (
    C1 = 8
    C2 = 13
    M = 15
)

func score(tip Tip) Score {
    
    // if the LSMI to YMRSI delta is over C1, then the tip is lazy
    if (LSMI - YMRSI(tip) > C1) {
        return Score.LAZY
    }
    
    // if the OMRSI to LSMI delta is over M/below-max-depth, then the tip is lazy
    if (LSMI - OMRSI(tip) > M) {
        return Score.LAZY
    }
    
    if (LSMI - OMRSI(tip) > C2) {
        return Score.SEMI_LAZY
    }

    return Score.NON_LAZY
}
```

### Random Tip-Selection

A node should keep a set of non-lazy tips (score 2).

> Every time a node is asked to select to tips to be approved it will just pick randomly from the set. 

Reviewer: ```suggestion
Every time a node is asked to select tips to be approved, it will pick randomly from the set. 
```

> A node must not execute tip-selection if it is not synchronized.
> 
> A tip should not be removed from the tips set immediately after it was selected in `select()`, 
> in order to make it possible for it to be re-selected, which in turn makes the Tangle wider
> and improves synchronization speed. A tip is removed from the tips set if `X` amount of direct
> approvers are reached or if a certain amount of time `T` passed. 
> It is recommended to use `X` = 2 and `T` = 3  but the threshold should be configurable.
> 
> ### Purpose Of Semi-Lazy Tips
> 
> Semi-Lazy tips are not eligible for tip-selection, but the coordinator node may implement a tip selection algorithm
> that confirms semi-lazy tips. Semi-lazy tips will usually be left behind, but parties interested in having them confirmed
> are incentivized to run spammers that will actively reduce the amount of semi-lazy tips eligible for coordinator's tip selection. 
> Given a coordinator that chooses semi-lazy tips, running such spammers may get those messages confirmed before
> they become lazy.
> 
> 
> # Drawbacks
> 
> Depending on when and how often `YMRSI`/`OMRSI` values are computed, this tip-selection could still
> have a slow runtime, as one would need to constantly walk down the Tangle in order to compute those
> values. However, smart caching might resolve this issue. 
> 
> # Rationale and alternatives
> 
> The previous tip-selection was written in accordance to the original IOTA whitepaper, as it also
> functioned as part of the consensus mechanism.
> However, relatively soon it became apparent that the cumulative weight computation was too heavy
> for an actual high throughput scenario and as such, the CW calculation is currently not used within
> node implementations at all.
> 
> Because confirmations with the [white-flag](https://github.com/iotaledger/protocol-rfcs/blob/master/text/0005-white-flag/0005-white-flag.md) approach no longer only approve cones with state mutations
> which are consistent with a previous ledger state, it makes sense to alter the tip-selection to provide 
> a fast way to get tips to approve with one's own message.
> The only important thing is to disincentive lazy behaviour in order to be able to maximize confirmation rate.

Reviewer: ```suggestion
A node must not execute tip-selection if it is not synchronized.

A tip should not be removed from the tips set immediately after it was selected in `select()`,  to make it possible for it to be re-selected, which in turn makes the Tangle wider
and improves synchronization speed. A tip is removed from the tips set if `X` amount of direct
approvers are reached or if a certain amount of time `T` passed. 
It is recommended to use `X` = 2 and `T` = 3  but the threshold should be configurable.

### Purpose Of Semi-Lazy Tips

Semi-Lazy tips are not eligible for tip-selection, but the coordinator node may implement a tip selection algorithm
that confirms semi-lazy tips. Semi-lazy tips will usually be left behind, but parties interested in having them confirmed
are incentivized to run spammers that will actively reduce the amount of semi-lazy tips eligible for coordinator's tip selection. 
Given a coordinator that chooses semi-lazy tips, running such spammers may get those messages confirmed before
they become lazy.

# Drawbacks

Depending on when and how often `YMRSI`/`OMRSI` values are computed, this tip-selection could still
have a slow runtime, as one would need to constantly walk down the Tangle to compute those
values. However, smart caching might resolve this issue. 

# Rationale and alternatives

The previous tip-selection was written in accordance to the original IOTA whitepaper, as it also
functioned as part of the consensus mechanism.
However, relatively soon it became apparent that the cumulative weight computation was too heavy
for an actual high throughput scenario and, as such, the CW calculation is currently not used within
node implementations at all.

Because confirmations with the [white-flag](https://github.com/iotaledger/protocol-rfcs/blob/master/text/0005-white-flag/0005-white-flag.md) approach no longer approve cones only with state mutations,
which are consistent with a previous ledger state, it makes sense to alter the tip-selection to provide 
a fast way to get tips to approve with one's own message.
The only important thing is to disincentive lazy behaviour to be able to maximize confirmation rate.
```

`Confirmed Root Transactions` of a given transaction.
* `Latest Solid Milestone Index (LSMI)` the index of the latest solid milestone.
* `Below Max Depth (BMD)` defines a threshold value up on which it is decided on whether a transaction is not
relevant in relation to the recent parts of the Tangle. The current `BMD` for mainnet nodes is 15 milestones, 
which means that transactions of which their `OTRSI` in relation to the `LSMI` is more than 15, are "below max depth".

### OTRSI / YTRSI example
Given the blue PoV transaction, the `OTRSI` of it is milestone 1 and `YTRSI` milestone 2.
Note that here again, the milestones are also `Confirmed Root Transactions`.
![sdf](images/otrsi_ytrsi.PNG)

### Milestone based tip scoring

The milestone based scoring defines a tip's score by investigating the tip's relation to
the cone it approves and previous issued  milestones.

A tip can have one of 3 score states:
* `0`: The tip is lazy and should not be selected.
* `1`: The tip is somewhat lazy.
* `2`: The tip is a non-lazy tip.

Definitions:
* `C1`: Max allowed delta value for the `YTRSI` of a given transaction in relation to the current `LSMI`.

> * `C2`: Max allowed delta value between `OTRSI` of the approvees of a given transaction in relation to the current `LSMI`. 

Reviewer 2: ```suggestion
* `C2`: Max allowed delta value between `OTRSI` of a given transaction in relation to the current `LSMI`. 
```
We don't calculate the OTRSI of the TX's approvees anymore, but of the TX itself.

    if (LSMI - OTRSI(tip) > M) {
        return Score.LAZY
    }
    
    if (LSMI - ORTSI(tip) > C2) {
        return Score.SEMI_LAZY
    }

    return Score.NON_LAZY
}
```

### Random Tip-Selection

A node should keep a set of non-lazy tips (score 2).
Every time a node is asked to select to tips to be approved it will just pick randomly from the set. 
A node must not execute tip-selection if it is not synchronized.

A tip should not be removed from the tips set immediately after it was selected in `select()`, 
in order to make it possible for it to be re-selected, which in turn makes the Tangle wider
and improves synchronization speed. A tip is removed from the tips set if `X` amount of direct
approvers are reached or if a certain amount of time `T` passed. 
It is recommended to use `X` = 2 and `T` = 3  but the threshold should be configurable.

### Purpose Of Semi-Lazy Tips

Semi-Lazy tips are not eligible for tip-selection, but the coordinator node may implement a tip selection algorithm
that confirms semi-lazy tips. Semi-lazy tips will usually be left behind, but parties interested in having them confirmed

> are incentivized to run spammers that will merge them into one thin subtangle. 

Reviewer 2: > will merge them into one thin subtangle. 

I find this sentence confusing: I would rephrase it in something like "... incentivized to run spammers that will actively reduce the amount of semi-lazy tips eligible for coordinator's tip selection"

Given a coordinator that chooses semi-lazy tips, running such spammers may get those transactions confirmed before
they become lazy.

# Drawbacks

Depending on when and how often `YTRSI`/`OTRSI` values are computed, this tip-selection could still
have a slow runtime, as one would need to constantly walk down the Tangle in order to compute those
values. However, smart caching might resolve this issue. 

# Rationale and alternatives

The previous tip-selection was written in accordance to the original IOTA whitepaper, as it also
functioned as the consensus mechanism to determine a transaction's confirmation rate.
However, relatively soon it became apparent that the cumulative weight computation was too heavy
for an actual high throughput scenario and as such, the CW calculation is currently not used within
node implementations at all.

> Because confirmations with the `white-flag` approach no longer only approve cones with state mutations

Reviewer 2: Link to WF for reference?
